flask/flask-common/src/main/java/net/corda/flask/common/LockFile.java

```diff
@@ -4,41 +4,36 @@
 import lombok.SneakyThrows;
 
 import java.io.Closeable;
-import java.io.IOException;
-import java.io.RandomAccessFile;
+import java.nio.channels.FileChannel;
 import java.nio.channels.FileLock;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.EnumSet;
 
 @RequiredArgsConstructor
 public class LockFile implements Closeable {
 
     private final FileLock lock;
 
-    private static RandomAccessFile createLockFile(Path path) throws IOException {
-        Files.createDirectories(path.getParent());
-        if (!Files.exists(path)) Files.createFile(path);
-        return new RandomAccessFile(path.toFile(), "rw");
-    }
-
     @SneakyThrows
     public static LockFile acquire(Path path, boolean shared) {
-        RandomAccessFile raf = createLockFile(path);
-        return new LockFile(raf.getChannel().lock(0L, Long.MAX_VALUE, shared));
+        Files.createDirectories(path.getParent());
+        FileChannel channel = FileChannel.open(path, EnumSet.of(StandardOpenOption.CREATE, StandardOpenOption.READ, StandardOpenOption.WRITE));
```

Comment:
These lines are common to both `acquire()` and `tryAcquire()` and so could be refactored into
```java
private static FileChannel createChannel(Path path)
```
I see that there are also `DELETE_ON_CLOSE` and `TRUNCATE` options :grinning:.

Reply:
I've noticed them as well... they could be useful but we do not want to always delete the lockfile on close, we currently use 2 lock files:
- one to lock the jar cache to prevent a process from removing from the cache jar files that another process is using (cache cleanup is performed at process startup)
- one as a parent-process heartbeat so that the child can notice that its creator has exited

For the first case we surely do not want the lockfile to be deleted when a process terminates as the same lockfile is shared between all the Flask processes (I assume the deletion would fail on Windows); for the second case we want to delete the lock file when the last between the parent and the child process exits and setting those flags means the child (that normally exits before the parent) tries to delete the lockfile while the parent process still holds it

Reply:
Fair enough, I thought you were deleting the lock files manually afterwards.

Reply:
I'm going to verify this anyway

Reply:
> we are in [the first case](https://github.com/corda/corda-gradle-plugins/blob/woggioni/flask-bugfix/flask/flask-launcher/src/main/java/net/corda/flask/launcher/Launcher.java#L165) and I'm wondering whether this is correct (I'm thinking about the case were you forcefully terminate the parent process sending SIGINT, the child is only killed by a shutdown hook at the end so it is still alive waiting to acquire the lock when then finally block runs, but that should be ok because I think that if you send SIGINT the finally block doesn't run at all)

I've just looked at that file more closely, and am now wondering why these lines:
```java
ZipEntry javaAgentsEntry = jar.getEntry(Flask.Constants.JAVA_AGENTS_FILE);
if(javaAgentsEntry != null) {
    try (InputStream inputStream = jar.getInputStream(jar.getEntry(Flask.Constants.JAVA_AGENTS_FILE))) {
```
aren't written as:
```java
ZipEntry javaAgentsEntry = jar.getEntry(Flask.Constants.JAVA_AGENTS_FILE);
if(javaAgentsEntry != null) {
    try (InputStream inputStream = jar.getInputStream(javaAgentsEntry)) {
```

Reply:
yeah, that's just silly... I'll address it

Reply:
> we are in the first case and I'm wondering whether this is correct (I'm thinking about the case were you forcefully 
> terminate the parent process sending SIGINT, the child is only killed by a shutdown hook at the end so it is still alive 
> waiting to acquire the lock when then finally block runs, but that should be ok because I think that if you send SIGINT the 
> finally block doesn't run at all)

I confirm the `finally` block isn't executed when the parent process is forcefully terminated

Reply:
Sending `SIGINT` to a JVM sounds like UNIX-specific behaviour to me. I'm not aware that Windows has an equivalent. My dim memories of ancient Windows programming suggest that [Control-C handling is only for console programs](https://docs.microsoft.com/en-us/windows/console/ctrl-c-and-ctrl-break-signals).

Reply:
SIGINT is unix-specific but Windows has something similar for console applications (like corda) when you use CTRL+C, I think they are called "events"

Reply:
> SIGINT is unix-specific but Windows has something similar for console applications (like corda) when you use CTRL+C, I think they are called "events"

I think that's what I already said. But AFAIK the Windows JVM is _not_ a console application. If it were, wouldn't it open a console when run?